cpp/dolfinx/fem/DirichletBC.cpp

```diff
@@ -50,55 +51,119 @@ dolfinx::MPI::Comm create_symmetric_comm(MPI_Comm comm)
+
+/// Find all DOFs on this process that has been detected on another process
+/// @param[in] comm A symmetric communicator based on the forward
+/// neighborhood communicator in the IndexMap
 /// @param[in] map The IndexMap with the dof layout
-/// @param[in] dofs_local The IndexMap with the dof layout
-/// @return List of local dofs with boundary conditions applied but
-///   detected by other processes. It may contain duplicate entries.
+/// @param[in] bs_map The block size of the index map, i.e. the dof array
+/// @param[in] bs The block size of the dof array
+/// @param[in] dofs_local List of degrees of freedom local to process
+/// (unrolled). It might contain indices not found on other processes
+/// @returns List of degrees of freedom that was found on the other processes
+/// that are in the local range (including ghosts)
 std::vector<std::int32_t>
-get_remote_bcs1(const common::IndexMap& map,
-                const std::vector<std::int32_t>& dofs_local)
+get_remote_dofs(MPI_Comm comm, const common::IndexMap& map, int bs_map, int bs,
+                const xtl::span<const std::int32_t>& dofs_local)
 {
-  dolfinx::MPI::Comm comm
-      = create_symmetric_comm(map.comm(common::IndexMap::Direction::forward));
+  assert(bs_map >= bs);
 
-  // Get number of processes in neighborhood
   int num_neighbors(-1), outdegree(-2), weighted(-1);
-  MPI_Dist_graph_neighbors_count(comm.comm(), &num_neighbors, &outdegree,
-                                 &weighted);
+  MPI_Dist_graph_neighbors_count(comm, &num_neighbors, &outdegree, &weighted);
+  assert(num_neighbors == outdegree);
 
   // Return early if there are no neighbors
   if (num_neighbors == 0)
     return {};
 
   // Figure out how many entries to receive from each neighbor
-  const int num_dofs = dofs_local.size();
+  assert(dofs_local.size() % bs == 0);
+  const int num_dofs_block = dofs_local.size() / bs;
   std::vector<int> num_dofs_recv(num_neighbors);
-  MPI_Neighbor_allgather(&num_dofs, 1, MPI_INT, num_dofs_recv.data(), 1,
-                         MPI_INT, comm.comm());
+  MPI_Request request;
+  MPI_Ineighbor_allgather(&num_dofs_block, 1, MPI_INT, num_dofs_recv.data(), 1,
+                          MPI_INT, comm, &request);
+
+  std::vector<std::int64_t> dofs_global(num_dofs_block);
+
+  // Get offset into map block
+  // NOTE: Could use neighbourhood allgather here. Also, not needed if
+  // bs_map == 1.
+  const int myoffset = dofs_local.empty() ? 0 : dofs_local.front() % bs_map;
+  int offset = 0;
+  MPI_Request request_offset;
+  MPI_Iallreduce(&myoffset, &offset, 1, MPI_INT, MPI_MAX, comm, &request_offset);
+
+  if (bs_map == 1)
+    map.local_to_global(dofs_local, dofs_global);
+  else
+  {
+    // Convert dofs indices to 'block' map indices
+    std::vector<std::int32_t> dofs_local_block;
+    dofs_local_block.reserve(num_dofs_block);
+    for (std::size_t i = 0; i < dofs_local.size(); i += bs)
+      dofs_local_block.push_back(dofs_local[i] / bs_map);
```

Comment:
@garth-wells here you have added the assumption that the dofs are sorted and contain no duplicates (which the input currently do) . Should we add that as a `@note for the input arguments`?

Reply:
> @garth-wells here you have added the assumption that the dofs are sorted and contain no duplicates 

Not sure that is entirely true on ordering. There is an assumption that blocks remained blocked.

> (which the input currently do) . Should we add that as a `@note for the input arguments`?

The docs already say the dofs must be sorted (in docs for in argument). Feel free to add more if you think it's needed.

Reply:
Agree that inside `get_remote_dofs` it does not have to be fully sorted, but the input has to be "blocked".
I will add a note on `get_remote_dofs` that we assume that the input dofs are blocked, but not necessarily sorted.